test/cmd/apply.sh

```diff
@@ -438,6 +438,165 @@ run_kubectl_server_side_apply_tests() {
+  name: test
+data:
+  key: value
+  newKey: newValue
+__EOF__
+  )
+
+  kube::test::get_object_assert "configmap test" "{{ .data.key }}" 'value'
+  kube::test::get_object_assert "configmap test" "{{ .data.newKey }}" 'newValue'
+  kube::test::get_object_assert "configmap test" "{{ .data.ssaKey }}" '<no value>'
+
+  # SSA the object without the field added above by CSA. Show that the object
+  # on the server has had the field removed
+  output_message=$(kubectl "${kube_flags[@]:?}" apply --server-side -f - << __EOF__
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+  ssaKey: ssaValue
+__EOF__
+  )
+
+  # Fetch the object and check to see that it does not have a field 'newKey'
+  kube::test::get_object_assert "configmap test" "{{ .data.key }}" 'value'
+  kube::test::get_object_assert "configmap test" "{{ .data.newKey }}" '<no value>'
+  kube::test::get_object_assert "configmap test" "{{ .data.ssaKey }}" 'ssaValue'
+
+  # Show that kubectl diff --server-side also functions after a migration
+  output_message=$(kubectl diff "${kube_flags[@]:?}" --server-side -f - << __EOF__ || test $? -eq 1
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+  annotations:
+    newAnnotation: newValue
+data:
+  key: value
+  newKey: newValue
+__EOF__
+)
+  kube::test::if_has_string "${output_message}" '+  newKey: newValue'
+  kube::test::if_has_string "${output_message}" '+    newAnnotation: newValue'
+
+  # clean-up
+  kubectl "${kube_flags[@]:?}" delete configmap test
+
+  ## Test to show that supplying a custom field manager to kubectl apply
+  # does not prevent migration from client-side-apply to server-side-apply
+  output_message=$(kubectl "${kube_flags[@]:?}" apply --server-side=false --field-manager=myfm  -f - << __EOF__
+apiVersion: v1
+data:
+ key: value1
+ legacy: value2
+kind: ConfigMap
+metadata:
+ name: ssa-test
+__EOF__
+)
+  kube::test::if_has_string "$output_message" "configmap/ssa-test created"
+  kube::test::get_object_assert "configmap ssa-test" "{{ .data.key }}" 'value1'
+
+  # show that after migrating to CSA with a custom field manager, the
+  # last-applied-annotation is present
+  grep -q kubectl.kubernetes.io/last-applied-configuration <<< "$(kubectl get configmap ssa-test -o yaml "${kube_flags[@]:?}")"
+
+  # Migrate to server-side-apply by applying the same object
+  output_message=$(kubectl "${kube_flags[@]:?}" apply --server-side=true --field-manager=myfm  -f - << __EOF__
+apiVersion: v1
+data:
+ key: value1
+ legacy: value2
+kind: ConfigMap
+metadata:
+ name: ssa-test
+__EOF__
+)
+  kube::test::if_has_string "$output_message" "configmap/ssa-test serverside-applied"
+  kube::test::get_object_assert "configmap ssa-test" "{{ .data.key }}" 'value1'
+
+  # show that after migrating to SSA with a custom field manager, the 
+  # last-applied-annotation is dropped
+  ! grep -q kubectl.kubernetes.io/last-applied-configuration <<< "$(kubectl get configmap ssa-test -o yaml "${kube_flags[@]:?}")" || exit 1
```

Comment:
The CI failure is a shellcheck error about this line:

```
In ./test/cmd/apply.sh line 581:
  ! grep -q kubectl.kubernetes.io/last-applied-configuration <<< "$(kubectl get configmap ssa-test -o yaml "${kube_flags[@]:?}")"
  ^-- SC2251: This ! is not on a condition and skips errexit. Use `&& exit 1` instead, or make sure $? is checked.

Did you mean: 
   grep -q kubectl.kubernetes.io/last-applied-configuration <<< "$(kubectl get configmap ssa-test -o yaml "${kube_flags[@]:?}")" && exit 1

For more information:
  https://www.shellcheck.net/wiki/SC2251 -- This ! is not on a condition and ...
```

Reply:
yup. I think I fixed this while your review was pending. the logic in the test is fixed